File: day02/day02.js
const fs = require('fs');
const input = String(fs.readFileSync('input.txt'));

let lines = input.split('\n').map(function (line) {
  return line.split(' ').map(function (index) {
    return Number(index);
  });
});

const authenticate = function (entry) {
  let currentEntry = [...entry];
  let currentEntryReversed = [...entry];
  currentEntry.sort(function (a, b) {
    return a - b;
  });
  currentEntryReversed.sort(function (a, b) {
    return b - a;
  });

  if (
    !(
      JSON.stringify(currentEntry) === JSON.stringify(entry) ||
      JSON.stringify(currentEntryReversed) === JSON.stringify(entry)
    )
  ) {
    return false;
  }

  for (let index = 0; index < entry.length - 1; index++) {
    if (Math.abs(entry[index] - entry[index + 1]) === 0) {
      return false;
    }
    if (Math.abs(entry[index] - entry[index + 1]) > 3) {
      return false;
    }
  }
  return true;
};

const test = lines.filter(authenticate);

console.log('Part 1:', test.length);

const removeOne = function (entry) {
  for (let index = 0; index < entry.length; index++) {
    let currentSplice = Array(entry).flat();
    currentSplice.splice(index, 1);
    if (authenticate(currentSplice)) {
      return true;
    }
  }
};

const test2 = lines.filter(removeOne);
console.log('Part 2:', test2.length);

File: day02/day02.py
lines = []

with open("input.txt", "r") as file:
    for line in file:
        lines.append([int(x) for x in line.strip().split()])

def authenticate(entry):
    current_entry_sorted = sorted(entry)
    current_entry_sorted_reversed = sorted(entry, reverse=True)

    if current_entry_sorted != entry and current_entry_sorted_reversed != entry:
        return False
    
    for index in range(len(entry) - 1):
        if abs(entry[index] - entry[index + 1]) == 0:
            return False
        elif abs(entry[index] - entry[index + 1]) > 3:
            return False
    return True

test = [x for x in lines if authenticate(x)]
print("Part 1:", len(test))

def remove_one(entry):
    for index in range(len(entry)):
        current_splice = list(entry)
        del current_splice[index]
        if authenticate(current_splice):
            return True

test_2 = ([x for x in lines if remove_one(x)])
print("Part 2:", len(test_2))